Full rewrite (#5)
pygfx/pylinalg: CONVENTIONS.md

# Style conventions

## Docstrings

Docstrings shall be written in NumpyDoc format.

> Docstrings can either be just the summary text, or they must be complete.

Author: I agree with @FirefoxMetzger's suggestion here.

Just a heads up; I'm not going to actually write any docstrings in this PR, I feel like the effort would likely be wasted due to changes later. I want to wait a bit until we have a little more stability.

I will still put the convention in place.

## Type annotations

Type annotations will not be applied (for now).

Note: I'm not very familiar with the latest and greatest type annotation

> options for libraries that make heavy use of numpy arrays like this one.

Author: Alright, that makes sense!

## Linting

Linting shall be performed with flake8, flake8-isort, flake8-black and pep8-naming.

Automated formatting shall be performed with black and isort.

Black is left at defaults, flake8 and isort are configured to adhere to black.

## Testing

All functions need to be covered by unit tests.

# Functional API conventions

This API is for internal use and for power-users that want to
vectorize operations on large sets of things.

Performance is prioritized over extensive input validation.

The source for this API resides in the `pylinalg.func` subpackage and is organized
by type.

## Function naming

The functional API has rather verbose names, but it makes things

> explicit.

Author: I intended to imply all the general code style rules by stating which linter plugins and setup we use. It includes `pep8-naming`.

I'll make sure to call it out explicitly.

## Function signatures

We strive to align closely with the conventions of [numpy](https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs)

> in order to be least-surprising for pylinalg users.

Author: Sort of. I'll incorporate this change slightly differently.

All functions shall adhere as much possible to the following signature:

> `func(x1, x2, ..., /, *, out=None, dtype=None, **kwargs)`
> 
> * All input arguments are positional-only.

Reviewer: That said, I would vote for extending this to 

```suggestion
`func(x1, x2, ..., /, *, out=None, dtype=None, **kwargs)`

* All input arguments are positional-only or keyword-only.
```

> * All input arguments are positional-only.

Author: Yes, my bad. I will provide clarification and an example would probably also help.

* There can be as many input arguments as needed for the function signature to make sense.

> * Input arguments are always [`array_like`](https://numpy.org/doc/stable/glossary.html?highlight=array_like#term-array_like):

Reviewer: I like the general idea behind this, but I fear it might be too restrictive. I can see scenarios where you want something like a `method="linear"` parameter or where you need a `eps=1e-6`. 

Of course, you can require that those are arrays and apply broadcasting to them to match the other inputs, but I am not sure if this adds value in comparison to making faulty inputs harder to catch.

Reviewer 2: IIUC Args like these would be keyword arguments. A bullet further down says "There may be additionaleyword-only arguments as required by the specifics of individual functions."

Author: Almar is correct. I tried to clarify this.

    * "Any scalar or sequence that can be interpreted as an ndarray. In
      addition to ndarrays and scalars this category includes lists
      (possibly nested and with different element types) and tuples.
      Any argument accepted by `np.array` is `array_like`."

>     * TBD: require a call to `np.asarray` or `np.asanyarray` on inputs?

Reviewer: I am against `np.asanyarray`; it allows object arrays and you usually don't want these in a numerics library. `np.asarray` or `np.ascontiguousarray` make sense in different places, but I wouldn't require it on each function.

For future reference, I like this SO answer as an overview: https://stackoverflow.com/a/52103839/6753182

Author: * I agree with `np.asarray` (and not `np.asanyarray`).
* Thanks for the SO reference. I know the differences, I'm just trying to reach consensus with you two on a decision. Thanks for the feedback, it helps make the right decision.

"`np.asarray` make sense in different places, but I wouldn't require it on each function." -> this is where the struggle lies; what is a good convention? When does it make sense, and when does it not make sense?

In my experience, it just gets added whenever people have the desire to call a function with a plain list instead of an ndarray, and the implementation just "accidentally" happens to perform an operation on that specific input that does not already implicitly call `np.asarray`.

Reviewer 2: FWIW `np.asarray` does not make a copy if the input already is an array. So it should not matter too much performance-wise.

Reviewer: > Thanks for the SO reference. I know the differences, I'm just trying to reach consensus with you two on a decision. Thanks for the feedback, it helps make the right decision.

Sorry, that may have come across differently from what I intended. I was thinking of this reference as a useful tool to discuss when we want to use which variant :)

> This is where the struggle lies; what is a good convention? When does it make sense, and when does it not make sense?

I agree with @almarklein here, that it is probably safe to add in most cases, since it is indeed a no-op if the array already is a base `ndarray` or a view and it is indeed an easy way to support ArrayLike input.

At the same time, there are two scenarios that come to mind where `np.asarray` is counter-productive. First, when it is important to preserve array subtypes, e.g. for sparse arrays, masked arrays, or if you have metadata (like ImageIOs old `image` class). Second, if you have an object whose buffer you could share with the base array, ie., where it is sufficient to create a view into a buffer instead of a dedicated array (I would have to check numpy's [`__array__` magic](https://numpy.org/devdocs/user/basics.dispatch.html) again to read up on if this can be solved elegantly).

Both cases can be handled separately though, so I would still opt for using `np.asarray` by default. That said, I would probably not make this a requirement in the conventions or at least state that there may be exceptions for special circumstances (like the above).

> * All non-input arguments are keyword-only.

Reviewer: Ah, I see.  In light of this, we can either ignore my previous comment on keyword-only arguments or revise this statement. I think I still prefer my approach because it's a bit unclear to me what a non-input argument is.

Author: I revised it.

* If `out` is `None` (the default), a new return array is created. Otherwise,
  the result of the function is written to `out`.
    * If there are multiple outputs, `out` is expected to be a tuple

>       with matching number of elements.

Reviewer: This is not related directly to the guide, but `out` on transform functions can create problems, e.g. if I do thinks like 

```
gaussian_blur(image, 0.2, (5, 5), out=image)
```

Are we worrying about such scenarios here, or will we cross that bridge when we get there?

Author: Unit tests will have to ensure that those cases are covered properly, I think.

Reviewer 2: @Korijn do you mean that the functions themselves will check for such cases and act appropriately? (i.e. not something the caller should worry about?)

Author: Yes

* If `dtype` is `None` (the default), the dtype of the return array
  will be determined automatically by numpy based on the operations
  performed on the inputs. Otherwise, the dtype of the output array(s) is
  overridden with the
  provided dtype with a call to `ndarray.astype`. This should ensure
  a matching precision of the calculation. The exact calculation
  dtypes chosen may depend on the function and the inputs may be cast

>   to this dtype to perform the calculation.

Author: I simplified this section. Essentially we just use numpy and don't bring anything else to the table. That means we default to numpy's default (float64) and you can override the data-types using the `out` and `dtype` arguments on your own if you want to.

* There may be additional keyword-only arguments as required by
  the specifics of individual functions.

Since the conventions align with those of numpy, in some cases, it is
possible to simply alias a numpy function to avoid all overhead and
implement the conventions. Optionally `functools.partial`
can be used to limit the available arguments, or a small wrapper function

> can be created.

Reviewer: @Korijn Do you have an example of when this would be preferable over just calling the corresponding numpy function?

Author: Sure. `matrix_new = partial(np.identity, 4)` creates a function that creates identity affine transform matrices, while the other arguments of the underlying numpy function are still available (e.g. `out` and `dtype`).

You could also write:

```py
def matrix_new(*args, **kwargs):
    return np.identity(4, *args, **kwargs)
``` 

But that is more code, and also significantly slower. If you have this code in a hot code path, the `partial` will perform much better.

Reviewer: What I was wondering is how `x = matrix_new()` improves `x=np.identity(4)`. Am I guessing correctly that it is meant for better readability?

Author: I guess to me it's similar to how it is can be easier for some users when everything they need is provided by `pygfx` and they don't need to also work with `wgpu`. :)

# Object oriented API conventions

This API is for external use and for novice-users that want to
perform linear algebra operations on conceptually familiar primitives.

In this API each "thing" is represented as one object. 
This API should make any linalg work much easier and safer, partly
because semantics matters here: a point is not the same as a vector.

> These objects are array-like and iterable to make them easy to

Reviewer: Does this mean that one `Point` object can hold multiple points, or that that the coordinates (x, y, z) of a point are iterable and indexable?

Author: The latter.

Reviewer: Got it. In that case, maybe we could say the following to be more precise?

```suggestion
These objects are backed by an array-like structure and are thus easy to
```

convert to native Python/Numpy objects. The objects support native python
operators such as `__mul__` where applicable, and have methods
specific to the type of object.

Extensive input validation and ease of use is prioritized over performance.

The source for this API resides in the `pylinalg.obj` subpackage and is organized
by type.

## Imports cycles

Since the classes here will regularly need to instantiate other types,
circular import dependencies exist. To work around this, only the `LinalgBase`
type can be imported at module level, and other types will have to be imported

> at runtime in methods.

Reviewer: This sounds a bit like our folder structure is suboptimal for what we aim to do. Is this an actual problem right now, or just a concern for the future?

Reviewer 2: I can imagine e.g. `Point.distance_to_plane()`  and `Plane.distance_to_point`, just to name one.

Author: Well, even simpler, `Point.to_vector()` and `Vector.to_point()` would be two functions that need to call each other's constructor. I think runtime imports are the proper way to deal with that.

Author: Why didn't I think of that? Will apply.

## Function naming

* Names should be concise and short.
* For every method, there is an alternative in-place method, signified
  with the prefix `i`. This is not the prettiest option, but it is

>   concise and short.

Reviewer: This seems contradictory to the aim of "Extensive input validation and ease of use is prioritized over performance.". 

There is no upside to in-place computation other than performance and if you get it wrong you end up with really hard-to-debug problems which is not very beginner friendly. I would vote for excluding in-place computation and reserving it for the functional/advanced API (which can do it via `method(x1, out=x1)`).

Author: There is another upside; element-specific mutation detection can be implemented efficiently only with in-place mutations.

I'm inclined to agree with you but I'm having a hard time seeing how pygfx's `WorldObject` attributes `position` etc would work when they would need to be replaced all the time.

Reviewer 2: Being able to assume that these objects are immutable also has favorable consequences (e.g. not having to worry about making a copy because the object might be modified without you knowing). But I agree that we need to see how this works out in PyGfx.

Reviewer: Fair point. Could there be a middle ground by saying that objects are mutable in general, but that methods return copies? This would allow us to do in-place operations via the functional API (e.g. using `out=`) giving us access to performance while preserving the "validated and safe" aim of the object API.

I also like this because it makes it explicit where you could run into trouble: You are switching to the functional API here, so you need to pay extra attention to what you are doing.

Author: Illustrative example, I think:

```py
camera.position.z = 250
```

Would become:

```py
camera.position = camera.position + [0, 0, 250]
```

I don't think that's an improvement from the user perspective.

    dtype : data-type, optional
        Overrides the data type of the result.

    Returns
    -------
    ndarray, [..., ndim]
        transformed vectors
    """
    vectors = vector_make_homogeneous(vectors, w=w)
    # usually when applying a transformation matrix to a vector
    # the vector is a column, so if you were to have an array of vectors
    # it would have shape (ndim, nvectors).
    # however, we instead have the convention (nvectors, ndim) where
    # vectors are rows.
    # therefore it is necessary to transpose the transformation matrix
    # additionally we slice off the last row of the matrix, since we are not interested
    # in the resulting w coordinate
    transform = matrix[:-1, :].T
    if out is not None:
        try:
            # if `out` is exactly compatible, that is the most performant
            return np.dot(vectors, transform, out=out)
        except ValueError:
            # otherwise we need a temporary array and cast
            out[:] = np.dot(vectors, transform)
            return out
    # otherwise just return whatever dot computes
    out = np.dot(vectors, transform)
    # cast if requested
    if dtype is not None:

>         out = out.astype(dtype)

Author: `copy=False`
